feat(VPie): create new component

J-Sek opened the pull request:
## Description

closes #20391

Includes
- `VPie` - main component
- `VPieSegment` - based on SVG
- `VPieTooltip` - floating VListItem with color indicator, title and value
- `useTransition` - simplified version with API similar to the equivalent from VueUse

TODO:
- props / API
  - [x] `series` » `items`
  - [x] `width` » `inner-cut`
  - [x] `legend` as complex prop `{ position, textFormat }`
  - [x] `animation` as complex prop `{ duration, easing }`
  - [x] `palette` prop
- misc
  - [x] new docs page with examples
  - [x] props descriptions
  - ~~tests ?~~ later
- internals
  - [x] `VTooltip` with `VAvatar` and `VListItem` as default
  - (later) [a11y] ~~SVG patterns built into the framework - similar to `decals` from ECharts~~
 
 Notes:
 - SVG rendering artifacts on semi-transparent slice (visible on the screenshot below) might depend on device and the browser (Chromium). Just checked and it looks fine on both Zen and Firefox

<details>
<summary>JS bundle size impact</summary>

| File                | Before    | After              |
| ------------------- | --------- | ------------------ |
| vuetify-labs.esm.js | 991.82 kB | 1.01 MB (up by 40~45 kB) |

</details>

## Markup:

<details><summary>Basic usage</summary>

```vue
<template>
  <v-app>
    <v-container>
      <div class="d-flex ga-12 align-start justify-center">
        <v-pie
          :items="items"
          :palette="palette"
          item-key="id"
          title="Basic Pie"
          animation
          legend
          tooltip
        />
        <v-pie
          :inner-cut="85"
          :items="items"
          :palette="palette"
          item-key="id"
          title="Funky donut"
          animation
          tooltip
        />
        <v-pie
          :hover-scale=".2"
          :inner-cut="70"
          :items="items"
          :palette="palette"
          item-key="id"
          title="Moooouuu..."
          animation
          hide-slice
          tooltip
        >
          <template #center>
            <v-icon color="#4e9963" icon="mdi-cow" size="90" />
          </template>
        </v-pie>
        <v-pie
          :gauge-cut="140"
          :hover-scale=".2"
          :inner-cut="50"
          :items="items"
          :palette="palette"
          item-key="id"
          title="Gauge"
          animation
          hide-slice
          tooltip
        />
      </div>
    </v-container>
  </v-app>
</template>

<script setup lang="ts">
  const palette = ['#2b6d40', '#4e9963', '#72c789', '#97f7b0']
  const items = [
    { id: 1, title: 'Series A', value: 45 },
    { id: 2, title: 'Series B', value: 30 },
    { id: 3, title: 'Series C', value: 15 },
    { id: 4, title: 'Series D', value: 10 },
  ]
</script>
```
</details>

<details><summary>Kitchen sink</summary>

```vue
<template>
  <v-app>
    <v-btn
      class="ma-2"
      icon="mdi-theme-light-dark"
      location="top right"
      position="absolute"
      @click="$vuetify.theme.toggle()"
    />
    <v-container>
      <div class="d-flex my-6 justify-center">
        <v-card class="pa-6" elevation="6" rounded="xl">
          <v-card-title class="d-flex align-center justify-space-between">
            <div class="text-truncate">Expense Analysis</div>
            <v-select
              v-model="expenseAnalysisGroup"
              :items="['Transactions', 'Other']"
              density="compact"
              max-width="200"
              variant="solo-filled"
              flat
              hide-details
              single-line
            />
          </v-card-title>
          <v-pie
            :key="expenseAnalysisGroup"
            :animation="{ duration: 250 }"
            :gap="2"
            :inner-cut="75"
            :items="items5"
            :legend="{ position: 'right' }"
            :rounded="2"
            :size="300"
            :tooltip="{ subtitleFormat: '[value]%' }"
            class="pa-3 mt-3"
            item-key="id"
            hide-slice
            reveal
          >
            <template #center>
              <div class="text-center">
                <div class="text-h3">130</div>
                <div class="opacity-70 mt-1 mb-n1">Total</div>
              </div>
            </template>
            <template #legend="{ items, toggle, isActive }">
              <v-list class="py-0 bg-transparent" density="compact" width="300">
                <v-list-item
                  v-for="item in items"
                  :key="item.key"
                  :class="['my-1', { 'opacity-40': !isActive(item) }]"
                  :title="item.title"
                  rounded="lg"
                  link
                  @click="toggle(item)"
                >
                  <template #prepend>
                    <v-avatar :color="item.color" :size="16" />
                  </template>
                  <template #append>
                    <div class="font-weight-bold">{{ item.value }}%</div>
                  </template>
                </v-list-item>
              </v-list>
            </template>
          </v-pie>
        </v-card>
      </div>

      <div class="d-flex ga-12 justify-center">
        <v-pie
          :hover-scale=".2"
          :items="items1"
          :size="160"
          density="compact"
          item-key="id"
          title="Basic Pie"
          reveal
        >
          <template #legend-text="{ item }"><b>{{ item.title }}</b>: {{ item.value }}%</template>
        </v-pie>
        <v-pie
          :animation="{ duration: 250 }"
          :inner-cut="80"
          :items="items2"
          :size="220"
          density="comfortable"
          item-key="id"
          title="Funky donut"
          reveal
        />
      </div>
      <div class="d-flex mt-6 justify-center">
        <v-sheet class="pa-6" elevation="6" rounded="xl">
          <v-pie
            :animation="{ duration: animationDuration }"
            :density="['compact', 'comfortable', 'default'][density]"
            :inner-cut="75"
            :items="items3"
            :legend="{ position: 'right' }"
            :size="300"
            :tooltip="{
              subtitleFormat: (s) => `${formatNumber(s.value)} respondents (${(100 * s.value / 50212).toFixed(1)}%)`,
              transition: { name: 'scroll-y-reverse-transition', duration: 150 }
            }"
            item-key="id"
            title="JavaScript alternatives"
            hide-slice
            reveal
          >
            <template #center="{ total }">
              <div class="text-center">
                <v-icon class="opacity-60" icon="mdi-vote-outline" size="44" />
                <div class="mt-2">{{ formatNumber(total) }} votes</div>
              </div>
            </template>
            <template #legend-text="{ item }">
              <div class="d-flex ga-6">
                <div>{{ item.title }}</div>
                <div class="ml-auto font-weight-bold">
                  {{ formatNumber(item.value) }}
                </div>
              </div>
            </template>
          </v-pie>
          <v-slider
            v-model="animationDuration"
            :ticks="{ 0: 'faster', 1000: 'slower' }"
            class="mb-n6 mt-6"
            label="animation speed"
            max="1000"
            min="0"
            show-ticks="always"
            step="100"
          />
          <v-slider
            v-model="density"
            :ticks="{ 0: 'compact', 1: 'comfortable', 2: 'default' }"
            class="mb-n3 mt-6"
            label="legend density"
            max="2"
            min="0"
            show-ticks="always"
            step="1"
          />
        </v-sheet>
      </div>

      <div class="h-0">
        <svg height="0" version="1.1" width="0" xmlns="http://www.w3.org/2000/svg">
          <!-- source: https://pattern.monster -->
          <defs>
            <pattern
              id="pattern-0"
              height="20"
              patternTransform="rotate(145) scale(.2)"
              patternUnits="userSpaceOnUse"
              width="20"
            >
              <path d="M0 10h20zm0 20h20zm0 20h20zm0 20h20z" fill="none" stroke="rgb(var(--v-theme-surface))" stroke-width="3" />
            </pattern>
            <pattern
              id="pattern-1"
              height="20"
              patternTransform="scale(.5)"
              patternUnits="userSpaceOnUse"
              width="40"
            >
              <path
                d="M40 0 20-10V0l20 10zm0 10L20 0v10l20 10zm0 10L20 10v10l20 10zM0 20l20-10v10L0 30zm0-10L20 0v10L0 20zM0 0l20-10V0L0 10z"
                fill="none"
                stroke="rgb(var(--v-theme-surface))"
              />
            </pattern>
            <pattern
              id="pattern-2"
              height="8"
              patternTransform="scale(.5)"
              patternUnits="userSpaceOnUse"
              width="70"
            >
              <path
                d="M-.02 22c8.373 0 11.938-4.695 16.32-9.662C20.785 7.258 25.728 2 35 2s14.215 5.258 18.7 10.338C58.082 17.305 61.647 22 70.02 22M-.02 14.002C8.353 14 11.918 9.306 16.3 4.339 20.785-.742 25.728-6 35-6S49.215-.742 53.7 4.339c4.382 4.967 7.947 9.661 16.32 9.664M70 6.004c-8.373-.001-11.918-4.698-16.3-9.665C49.215-8.742 44.272-14 35-14S20.785-8.742 16.3-3.661C11.918 1.306 8.353 6-.02 6.002"
                fill="none"
                stroke="rgb(var(--v-theme-surface))"
              />
            </pattern>
            <pattern
              id="pattern-3"
              height="10"
              patternTransform="scale(.5)"
              patternUnits="userSpaceOnUse"
              width="10"
            >
              <path
                d="M5 0v10ZM0 5h10Z"
                fill="none"
                stroke="rgb(var(--v-theme-surface))"
              />
            </pattern>
          </defs>
        </svg>
      </div>
      <div class="d-flex mt-6 justify-center">
        <v-pie
          :hover-scale=".1"
          :inner-cut="50"
          :items="items4"
          :legend="{ position: 'left' }"
          :size="200"
          :tooltip="{ subtitleFormat: '[value]%' }"
          item-key="id"
          hide-slice
          segment
        >
          <template #center>
            <div class="text-center">
              <v-icon class="opacity-60" icon="mdi-leaf" size="44" />
            </div>
          </template>
        </v-pie>
      </div>
    </v-container>
  </v-app>
</template>

<script setup lang="ts">
  import { shallowRef, toRef } from 'vue'
  import { useTheme } from 'vuetify'

  const { change } = useTheme()
  change('dark')

  const numberFormatter = new Intl.NumberFormat('en', { useGrouping: true })
  function formatNumber (v: number) {
    return numberFormatter.format(v)
  }

  const expenseAnalysisGroup = shallowRef('Transactions')

  const items1 = [
    { id: 1, title: 'Series A', value: 45, color: '#2b6d40' },
    { id: 2, title: 'Series B', value: 30, color: '#4e9963' },
    { id: 3, title: 'Series C', value: 15, color: '#72c789' },
    { id: 4, title: 'Series D', value: 10, color: '#97f7b0' },
  ]

  const items2 = [
    { id: 2, title: 'Google', value: 75, color: '#0080bb' },
    { id: 1, title: 'Bing', value: 20, color: '#58508d' },
    { id: 3, title: 'DuckDuckGo', value: 17, color: '#bc5090' },
    { id: 4, title: 'Brave', value: 15, color: '#ff6361' },
    { id: 5, title: 'Kagi', value: 5, color: '#ffa600' },
  ]

  const items3 = [
    { id: 1, title: 'TypeScript', value: 17674, color: '#13475c' },
    { id: 2, title: 'Elm', value: 3550, color: '#006c71' },
    { id: 3, title: 'CoffeeScript', value: 1251, color: '#008e59' },
    { id: 4, title: 'Civet', value: 531, color: '#ffa600' },
    { id: 5, title: 'N/A', value: 9821, color: '#6662' },
  ]
  const animationDuration = shallowRef(250)
  const density = shallowRef(2)

  const items4 = [
    { id: 1, title: 'Walnut', value: 57, color: '#607322', pattern: 'url(#pattern-1)' },
    { id: 2, title: 'Oak', value: 31, color: '#c19a00', pattern: 'url(#pattern-2)' },
    { id: 3, title: 'Pine', value: 12, color: '#ffa600', pattern: 'url(#pattern-3)' },
  ]

  const items5 = toRef(() => expenseAnalysisGroup.value === 'Transactions'
    ? [
      { id: 1, title: 'House & Bills', value: 40, color: 'rgba(var(--v-theme-on-surface), .2)', pattern: 'url(#pattern-0)' },
      { id: 2, title: 'Transportation', value: 25, color: 'rgba(255, 151, 215, .4)' },
      { id: 3, title: 'Entertainment', value: 20, color: 'rgba(255, 151, 215, .6)' },
      { id: 4, title: 'Food', value: 10, color: 'rgba(255, 151, 215, .8)' },
      { id: 5, title: 'Other', value: 5, color: 'rgba(255, 151, 215, 1)' },
    ]
    : [
      { id: 1, title: 'OSS Donations', value: 37, color: '#767119' },
      { id: 2, title: 'Travel', value: 22, color: '#9e850d' },
      { id: 3, title: 'Investment', value: 20, color: '#cb9700' },
      { id: 4, title: 'Books', value: 11, color: '#ffa600' },
    ])
</script>
```
</details>

<details>
<summary>Gauge demo</summary>

```vue
<template>
  <v-app>
    <v-btn
      class="ma-2"
      icon="mdi-theme-light-dark"
      location="top right"
      position="absolute"
      @click="themeToggle()"
    />
    <v-container>
      <div class="d-flex my-6 justify-center">
        <v-card class="pa-6" elevation="6" rounded="xl">
          <v-pie
            :gap="2"
            :gauge-cut="gaugeCut"
            :items="items"
            :legend="{ position: 'right', textFormat: '[title] ([value]%)' }"
            :tooltip="{ subtitleFormat: '[value]%' }"
            class="pa-3 mt-n1"
            inner-cut="75"
            rounded="2"
            size="300"
            title="Demographics"
            animation
            hide-slice
          >
            <template #center>
              <div class="text-center">
                <div class="text-h3">890</div>
                <div class="opacity-70 mt-1 mb-n1">Participants</div>
              </div>
            </template>
          </v-pie>
        </v-card>
      </div>

      <div class="mx-auto" style="width: 400px">
        <v-card class="pa-6" elevation="6" rounded="xl">
          <v-slider v-model="gaugeCut" max="180" hide-details />
        </v-card>
      </div>
    </v-container>
  </v-app>
</template>

<script setup lang="ts">
  import { ref } from 'vue'
  import { useTheme } from '@/composables'

  const { change, toggle: themeToggle } = useTheme()
  change('dark')

  const gaugeCut = ref(120)

  const items = [
    { key: 1, title: 'Age: 18 ~ 29', value: 35, color: '#5c3047' },
    { key: 2, title: 'Age: 30 ~ 39', value: 30, color: '#c8534e' },
    { key: 3, title: 'Age: 40 ~ 49', value: 20, color: '#ee7739' },
    { key: 4, title: 'Age: 50+', value: 15, color: '#ffa600' },
  ]
</script>

<style>
  @import 'https://fonts.bunny.net/css?family=recursive:300,400,500,700';
  body * {
    font-family: 'Recursive', sans-serif !important;
  }
</style>
```

</details>

![image](https://github.com/user-attachments/assets/db2e3989-7efe-4a63-a496-3dfbc26b2acd)

![image](https://github.com/user-attachments/assets/7d8886d0-2d59-464e-8269-52d3b539368b)

Revadike commented:
I'd love this to be merged. What's the main thing that's holding it back?

J-Sek commented:
> I'd love this to be merged. What's the main thing that's holding it back?

Some cleanup, naming adjustments, bundle impact measurement. I will try to push it over the weekend.

Revadike commented:
> > I'd love this to be merged. What's the main thing that's holding it back?
> 
> Some cleanup, naming adjustments, bundle impact measurement. I will try to push it over the weekend.

Had time to work on it yet? 🤞

johnleider commented on packages/vuetify/src/labs/VPie/VPieSegment.tsx:
```diff
@@ -0,0 +1,144 @@
+  },
+  value: {
+    type: Number,
+    required: true,
```
no required values please

johnleider reviewed (changes requested):
Very extensive, love it. Don't take my feedback as nit picking.

J-Sek commented:
Rebased, added `reveal`, replaced `circle` with another `path` to simplify and make it more cohesive.
There is just a tiny fix in `locationStrategies.ts`, so maybe it could land in `master`?
or merged right away to be included in 3.9.0?

J-Sek commented:
Great news! I have just included a final fix. Centered content with `gauge-cut` was slightly misaligned.
Now... I'd hate to interrupt Kael fixing `dev` to pass through CI/CD. It can wait another day or two.

johnleider commented on packages/vuetify/src/components/VOverlay/locationStrategies.ts:
```diff
@@ -241,6 +241,11 @@ function connectedLocationStrategy (data: LocationStrategyData, props: StrategyP
   })
 
   watch([data.target, data.contentEl], ([newTarget, newContentEl], [oldTarget, oldContentEl]) => {
+    if (newContentEl === oldContentEl && Array.isArray(newTarget)) {
```
@KaelWD probably needs to look this over

KaelWD replied:
[`6cb4874` (#15682)](https://github.com/vuetifyjs/vuetify/pull/15682/commits/6cb4874618509cd31e5f571185e89cb2a07e0d14)

johnleider commented on packages/vuetify/src/composables/reveal.ts:
```diff
@@ -0,0 +1,54 @@
+      isRevealing.value = false
+      isRevealed.value = true
+    }
+  })
```
```suggestion
  })

```

KaelWD commented on packages/vuetify/src/labs/VPie/VPie.tsx:
```diff
@@ -0,0 +1,378 @@
+    }
+
+    return () => {
+      const itemProps = VPieSegment.filterProps(pick(props, [
```
Why filterProps if they're already picked? There's only a few, why not explicit passthrough like you have for innerCut and hoverScale?

KaelWD commented on packages/vuetify/src/labs/VPie/VPie.tsx:
```diff
@@ -0,0 +1,378 @@
+          <div
+            class={[
+              'v-pie__content',
+              ...textColorClasses.value,
```
```suggestion
              textColorClasses.value,
```

J-Sek replied:
Actually, this `textColorClasses` was always empty. Fixed by using background class and overriding background to none.

KaelWD commented on packages/vuetify/src/labs/VPie/VPie.tsx:
```diff
@@ -0,0 +1,378 @@
+              'v-pie__content',
+              ...textColorClasses.value,
+            ]}
+            style={{
+              transform: `rotate(${rotateDeg.value})`,
+              marginBottom: `calc(-1 * ${convertToUnit(props.size)} * ${gaugeOffset.value})`,
+              ...textColorStyles.value,
+            }}
```
```suggestion
            style={[
              {
                transform: `rotate(${rotateDeg.value})`,
                marginBottom: `calc(-1 * ${convertToUnit(props.size)} * ${gaugeOffset.value})`,
              },
              textColorStyles.value,
            ]}
```

KaelWD commented on packages/vuetify/src/util/easing.ts:
```diff
@@ -1,3 +1,74 @@
+    transition: easingPatterns.easeInOutCubic,
+  }
+
+  const raf: ReturnType<typeof requestAnimationFrame> = null!
```
I always use -1 for the initial value, then you don't need any assertions

J-Sek commented:
- items support `color`, palette is not required
- colors from theme [mostly supported](https://play.vuetifyjs.com/#eNp9k92OmzAQhV9lxE1ABUI2SreKkpWqXvcJNnvhmiG4a2zkH7pplHfvYAiwbbU34Bmf8+E5wPM1soavv7Zt3nmM9tHBYdNK5vDppAAOXca1ckwoNMAls/Z4isqskvgG/YV2pW8UMCnOKuOoHOl+eutEdRnLUxRIgWVrRDdxWpZ9PkWAEjvmhFbU6mujvSqxpOpNTuZgbwXeK4C9oJP2mHA/RYsdiWdUPWBYfNOqEuelQiiaJ+PekebLbrljxW+k5rYoll2mRBOOOLdqUWJmpeDTkdb3OdfjoEOCVE0RUuewXgRMpeVGtA4sOt8Gg2habRxcgeum9Q7LFAxWKfxijtdwg8roBlb0sla9H0imrINh0u+6RDj2+nhlxLl2q+RvyRDGKLpCq63oB9vDaIDbwhKiJe3zMNkVXvGyh4cUnHASydMaCsZcVil0THrqPO5S8kpteuD5R9wxE2dZl7kaG8xGeZLQY9J3zM3MtJ5ztHZmPhQfMUf5f5jbmYnGaDMTN48zUb9KXsch1X/hwZckwJnksYRPkG+ogHp62MscltOOSQrr/t7iOIHj05BhbrD0HOPY+iYFFTZoSUCVhzOlUCQU/GE9fA/0JUS3NNrmu3yzjfrFpsiLKK2YtJi++2WH3ssfUAEybQ==), just the VChip chokes on OKLCH transformation
